esda/lee.py

```diff
@@ -0,0 +1,143 @@
+            self.significance_ = (larger + 1.) / (self.permutations + 1.)
+        return self
+
+    @staticmethod
+    def __statistic(Z,W):
+        ctc = W.T @ W
+        ones = numpy.ones(ctc.shape[0])
+        return (Z.T @ ctc @ Z) / (ones.T @ ctc @ ones)
+
+
+class Local_Spatial_Pearson(BaseEstimator):
+    def __init__(self, connectivity = None, permutations=999):
+        self.connectivity = connectivity
+        self.permutations = permutations
+
+    def fit(self, x, y):
+        """
+        bivariate local pearson's R based on Eq. 22 in Lee (2001), using 
+        site-wise conditional randomization from Moran_Local_BV.
+        """
+        x = utils.check_array(x)
+        y = utils.check_array(y)
+        #x = preprocessing.StandardScaler().fit_transform(x)
+        #y = preprocessing.StandardScaler().fit_transform(y)
+        Z = numpy.column_stack((preprocessing.StandardScaler().fit_transform(x),
+                                preprocessing.StandardScaler().fit_transform(y)))
+        
+        #slx = self.connectivity @ x
+        #sly = self.connectivity @ y
+        n, _ = x.shape
+        #self.associations_ = n * (slx - x.mean()) * (sly - y.mean())
+        #self.associations_ /= x.std() * y.std() # should always be one
+
+        self.associations_ = self.__statistic(Z,self.connectivity)
+
+        if self.permutations:
+            W = WSP(self.connectivity).to_W()
+            self.reference_distribution_ = numpy.empty((n, self.permutations))
+            k = W.max_neighbors + 1
+            random_ids = numpy.array([numpy.random.permutation(n - 1)[0:W.max_neighbors + 1]
+                                      for i in range(self.permutations)])
+            ids = numpy.arange(n)
+            weights = [W.weights[idx] for idx in W.id_order]
+            cardinalities = [W.cardinalities[idx] for idx in W.id_order]
+
+            for i in range(n):
+                ids_not_i = ids[ids != i]
+                numpy.random.shuffle(ids_not_i)
+                randomizer = random_ids[:, 0:cardinalities[i]]
+                random_neighbors = ids_not_i[randomizer]
+                random_neighbor_x = x[random_neighbors]
+                random_neighbor_y = y[random_neighbors]
+
+                self.reference_distribution_[i]  = ((random_neighbor_x * weights[i]).sum() - x.mean())
+                self.reference_distribution_[i] *= ((random_neighbor_y * weights[i]).sum() - y.mean())
+            self.reference_distribution_ *= (n / x.std() * y.std())
+        return self
+
+    @staticmethod
+    def __statistic(Z,W):
+        ctc = W.T @ W
+        ones = numpy.ones(ctc.shape[0])
+        return (Z[:,1] @ W.T) * (W @ Z[:,0]) 
+
+if __name__ == '__main__':
+    import geopandas
+    import pysal
+    df = geopandas.read_file(pysal.examples.get_path('columbus.shp'))
+    x = df[['HOVAL']].values
+    y = df[['CRIME']].values
+    zx = preprocessing.StandardScaler().fit_transform(x)
```

Comment:
Should Lines 137 and 138 be removed? What's happening here seems to be happening in Line 90.

Reply:
That `if __name__ == '__main__':` part only runs if the python script itself is run. it helps when I'm exploring/writing the code, and it's not executed when the module is imported. I was standardizing these cause, when I run `lee.py` for troubleshooting, it's helpful to have `zx` and `zy`.